ADAL/src/validation/ADAadAuthorityCache.m

```diff
@@ -0,0 +1,215 @@
+              oauthError:(ADAuthenticationError *)oauthError
+                 context:(id<ADRequestContext>)context
+{
+    [self getWriteLock];
+    AD_LOG_WARN(@"Caching Invalid AAD Instance", context.correlationId, nil);
+    __auto_type *record = [ADAadAuthorityCacheRecord new];
+    record.validated = NO;
+    record.error = oauthError;
+    _map[authority.adHostWithPortIfNecessary] = record;
+    pthread_rwlock_unlock(&_rwLock);
+}
+
+#pragma mark -
+#pragma mark Cache Accessors
+
+- (ADAadAuthorityCacheRecord *)checkCacheImpl:(NSURL *)authority
+{
+    __auto_type record = _map[authority.adHostWithPortIfNecessary];
+    pthread_rwlock_unlock(&_rwLock);
+    
+    return record;
+}
+
+- (ADAadAuthorityCacheRecord *)tryCheckCache:(NSURL *)authority
+{
+    if (pthread_rwlock_tryrdlock(&_rwLock) == 0)
+    {
+        return [self checkCacheImpl:authority];
+    }
+    
+    return nil;
+}
+
+- (ADAadAuthorityCacheRecord *)checkCache:(NSURL *)authority
+{
+    int status = pthread_rwlock_rdlock(&_rwLock);
+    // This should be an extremely rare condition, and typically only happens if something
+    // (a memory stomper bug) stomps on the rw lock. In that case we're in a really bad state anyways
+    // and should expect to fail soon.
+    if (status != 0)
+    {
+        AD_LOG_ERROR(@"Failed to grab authority cache read lock.", status, nil, nil);
+        @throw [NSException exceptionWithName:@"ADALException"
+                                       reason:[NSString stringWithFormat:@"Unable to get lock, error code %d", status]
+                                     userInfo:nil];
+    }
+    
+    return [self checkCacheImpl:authority];
+}
+
+
+- (BOOL)getWriteLock
+{
+    int status = pthread_rwlock_wrlock(&_rwLock);
+    if (status != 0)
+    {
+        AD_LOG_ERROR(@"Failed to grab authority cache write lock.", status, nil, nil);
+        @throw [NSException exceptionWithName:@"ADALException"
+                                       reason:[NSString stringWithFormat:@"Unable to get lock, error code %d", status]
+                                     userInfo:nil];
+    }
+    
+    return YES;
+}
+
+static NSURL *urlForPreferredHost(NSURL *url, NSString *preferredHost)
+{
+    if (!preferredHost)
+    {
+        return url;
+    }
+    
+    // If the host is already (functionally) the same then just return the passed in URL.
+    if ([url.host isEqualToString:preferredHost] ||
+        [url.adHostWithPortIfNecessary isEqualToString:preferredHost])
+    {
+        return url;
+    }
+    
+    // Otherwise switch the host for the preferred one.
+    NSURLComponents *components = [NSURLComponents componentsWithURL:url resolvingAgainstBaseURL:NO];
+    
+    // I hope there's never a case where there's percent encoded characters in the host, but using
+    // this setter prevents NSURLComponents from trying to do any further mangling on the string,
+    // probably a good thing.
+    components.percentEncodedHost = preferredHost;
```

Comment:
From documentation: If you set this value to something that is not a valid, percent-encoded string, this class throws an exception. Is it sure preferredHost will be always a valid string?